wgengine/magicsock/magicsock.go

```diff
@@ -2555,11 +2558,72 @@ func (c *Conn) SetProbeUDPLifetime(v bool) {
 
 func capVerIsRelayCapable(version tailcfg.CapabilityVersion) bool {
 	// TODO(jwhited): implement once capVer is bumped
-	return false
+	return version == math.MinInt32
 }
 
+func capVerIsRelayServerCapable(version tailcfg.CapabilityVersion) bool {
+	// TODO(jwhited): implement once capVer is bumped
+	return version == math.MinInt32
+}
+
+// onFilterUpdate is called when a [FilterUpdate] is received over the
+// [eventbus.Bus].
 func (c *Conn) onFilterUpdate(f FilterUpdate) {
-	// TODO(jwhited): implement
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.filt = f.Filter
+	if !c.relayClientEnabled {
+		// Early return if we cannot operate as a relay client.
+		return
+	}
+	// The filter has changed, re-evaluate it in order to produce an updated
+	// relay server set.
+	relayServers := make(set.Set[netip.AddrPort])
+	for _, peer := range c.peers.All() {
+		peerAPI := peerAPIIfCandidateRelayServer(c.filt, c.self, peer)
+		if peerAPI.IsValid() {
+			relayServers.Add(peerAPI)
+		}
+	}
+	c.relayManager.handleRelayServersSet(relayServers)
+}
+
+// peerAPIIfCandidateRelayServer returns the peer API address of peer if it
+// is considered to be a candidate relay server upon evaluation against filt and
+// self, otherwise it returns a zero value.
+func peerAPIIfCandidateRelayServer(filt *filter.Filter, self, peer tailcfg.NodeView) netip.AddrPort {
```

Comment:
We should consider an algorithm that flips peer and cap matches iteration. If we iterate all caps matches and find the cap of interest we can then look for a matching peer. c.peers is a slice, but it's ordered by NodeID, so could be binary searched. We could also just make it a map.

Practically speaking this may be overthinking the time complexity and we probably don't have the right identifiers today to inverse match, but warrants a TODO consideration at least.

Reply:
Yep, we need addresses to eval in the filter. I'll probably just start with some TODOs on this func.